File: app/src/main/java/com/happs/ximand/ringcontrol/view/fragment/SelectDeviceFragment.java
package com.happs.ximand.ringcontrol.view.fragment;

import android.bluetooth.BluetoothDevice;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.happs.ximand.ringcontrol.R;
import com.happs.ximand.ringcontrol.databinding.FragmentSelectDevicesBinding;
import com.happs.ximand.ringcontrol.view.adapter.BluetoothDevicesRecyclerViewAdapter;
import com.happs.ximand.ringcontrol.viewmodel.fragment.SelectDeviceViewModel;

import java.util.List;

public class SelectDeviceFragment extends BaseFragmentWithRecyclerView<SelectDeviceViewModel,
        FragmentSelectDevicesBinding, BluetoothDevice, BluetoothDevicesRecyclerViewAdapter> {

    public SelectDeviceFragment() {
        super(R.layout.fragment_select_devices, 0);
    }

    public static SelectDeviceFragment newInstance() {
        return new SelectDeviceFragment();
    }

    @Override
    protected RecyclerView getRecyclerViewFromBinding(FragmentSelectDevicesBinding binding) {
        return binding.devicesRecyclerView;
    }

    @Override
    protected void onPreAttachRecyclerViewAdapter(BluetoothDevicesRecyclerViewAdapter adapter) {
        adapter.setDeviceSelectedListener(getViewModel()::notifyDeviceSelected);
    }

    @Override
    protected void onPreViewModelAttaching(@NonNull SelectDeviceViewModel viewModel) {
        viewModel.getDevicesLiveData().observe(getViewLifecycleOwner(), this::initAdapter);
        viewModel.getStartSettingsActivityLiveEvent().observe(
                getViewLifecycleOwner(), this::startSettingsActivity
        );
    }

    private void startSettingsActivity(Void aVoid) {
        Intent intentOpenBluetoothSettings = new Intent();
        intentOpenBluetoothSettings.setAction(android.provider.Settings.ACTION_BLUETOOTH_SETTINGS);
        startActivity(intentOpenBluetoothSettings);
    }

    @Override
    protected BluetoothDevicesRecyclerViewAdapter createNewAdapter(List<BluetoothDevice> items) {
        return new BluetoothDevicesRecyclerViewAdapter(items);
    }
}

File: app/src/main/java/com/happs/ximand/ringcontrol/view/fragment/SelectDeviceFragment.kt
package com.happs.ximand.ringcontrol.view.fragment

import android.bluetooth.BluetoothDevice
import androidx.appcompat.app.ActionBar
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.happs.ximand.ringcontrol.R
import com.happs.ximand.ringcontrol.databinding.FragmentSelectDevicesBinding
import com.happs.ximand.ringcontrol.databinding.ItemDeviceBinding
import com.happs.ximand.ringcontrol.view.adapter.BluetoothDevicesRecyclerViewAdapter
import com.happs.ximand.ringcontrol.viewmodel.fragment.SelectDeviceViewModel

class SelectDeviceFragment : BaseFragmentWithRecyclerView<SelectDeviceViewModel,
        FragmentSelectDevicesBinding, BluetoothDevice, ItemDeviceBinding, BluetoothDevicesRecyclerViewAdapter>
(R.layout.fragment_select_devices, MENU_NONE) {

    companion object {
        @JvmStatic
        fun newInstance(): SelectDeviceFragment {
            return SelectDeviceFragment()
        }
    }

    private var swipeRefreshLayout: SwipeRefreshLayout? = null

    override fun onSetActionBarTitle(actionBar: ActionBar) {
        actionBar.setTitle(R.string.connect)
    }

    override fun onViewDataBindingCreated(binding: FragmentSelectDevicesBinding) {
        super.onViewDataBindingCreated(binding)
        swipeRefreshLayout = binding.swipeRefreshLayout
        swipeRefreshLayout!!.setOnRefreshListener {
            onRefresh()
        }
    }

    private fun onRefresh() {
        requireViewModel().updateData(requireContext()) {
            swipeRefreshLayout!!.isRefreshing = false
        }
    }

    override fun getRecyclerViewFromBinding(binding: FragmentSelectDevicesBinding): RecyclerView {
        return binding.devicesRecyclerView
    }

    override fun onPreAttachRecyclerViewAdapter(adapter: BluetoothDevicesRecyclerViewAdapter) {
        adapter.deviceSelectedListener = requireViewModel()::connectToDevice
        adapter.connecting = requireViewModel().connecting
    }

    override fun onPreViewModelAttaching(viewModel: SelectDeviceViewModel) {
        viewModel.devicesLiveData.observe(viewLifecycleOwner, Observer {
            initAdapter(it)
        })
    }

    override fun onResume() {
        super.onResume()
        onRefresh()
    }

    override fun createNewAdapter(items: MutableList<BluetoothDevice>): BluetoothDevicesRecyclerViewAdapter {
        return BluetoothDevicesRecyclerViewAdapter(items)
    }
}